### DISCUSSION TITLE
Need some pointers on using @xstate/test with complex machines

### DISCUSSION THREAD

**Discussion by AlCalzone** [statelyai/xstate#1464] [General]
As of now, I have a pretty complicated state machine, which invokes a relatively simple one a couple of times during its execution.
My current testing setup is a huge bunch of single tests which place the machine(s) in specific states, send it events and assert that some stuff changed or some external methods were called. You can see that in action at https://github.com/AlCalzone/node-zwave-js/blob/73020fd2569687a47af592c7aed5ccdc1fbe5808/packages/zwave-js/src/lib/driver/SendThreadMachine.test.ts#L158

This works, but has some huge flaws:
* It is a shitload of test code
* It certainly feels hacky
* It is very brittle because the tests perform assertions on the internal state of the machine

It turns out that I missed something when creating said machine, so I'm in the process of refactoring it. I was thinking of using two machines that communicate using events. However, this also means that I need to update all my tests. After watching the talk on model based testing with @xstate/test, I feel that is exactly what @xstate/test was made for, but I cannot yet make sense on how to use it for my problem.

A few requirements:
* This is all "backend" code, no UI
* I'm using jest for testing
* The machine has no final state because it active during the entire lifetime of my application
* Some transitions have side effects and I need to assert that these side effects are executed
* The flow has loops (some actions are retried up to 3 times)
* The tests must be able to handle timeouts

**Comment by rjdestigter** [comment]
Make sure you differentiate between the machine configuration used to model your test from the machine used in your application. The second _is the system under test (SUT)_.

**Reply by karfau** [reply]
Would this also hold true when developing a "machine factory" function (that encapsulates creating a machine with certain behavior)?

```
export const createXMachine = (options) => {
  return Machine(
    // ... use options to configure a machine
  )
}
```

So far I was executing the test plans of that machine, but I'm struggling with asserting that certain actions have been applied.

Update: I managed to write a test manually that `interpret`s the machine and asserts on states and context value. But I wonder if that should be also possible with model based testing.

(I can also create a new thread, but this one seemed to be a nice fit.)

**Comment by rjdestigter** [comment]
> Some transitions have side effects and I need to assert that these side effects are executed

Reading https://dev.to/rjdestigter/working-with-non-user-asynchronous-events-in-model-based-tests-with-xstate-58ek might be helpful if your looking at testing asynchronous effects.

If an action on transition performs a synchronous side effect but you don't have access to any testable information that would indicate that the effect was performed then you could override the machine configuration when intepreting it  your test:

```ts
// machine.ts
const machine = createMachine({
    ...
    states: {
        foo: {
            entry: 'myNamedSideEffect
        }
    }
}, {
    actions: {
        myNamedSideEffect: () => {
            sideEffect();
        }
    }
})

// machine.test.ts

import { interpret } from 'xstate'
import machine from './'machine'

let ranSideEffect = false;

describe('Testing my machine', () => {
    const service = interpret(machine, {
        actions: {
            myNamedSideEffect: [
                () => { ranSideEffect = true },
                machine.configuration.actions.myNamedSideEffect
            ]
        }
    })
})
```
This would allow you to test if _ranSideEffect_ was mutated.

**Reply by AlCalzone** [reply]
Thanks for the links, reading different approaches certainly helps understand how this all comes together.

**Comment by AlCalzone** [comment]
Ok, I'm starting to realize that I can't just take my existing machine and generate tests for it. I have to design a test machine that interacts with the machine that should be tested. Until now I've only found examples where some UI testing is done, but I can't yet wrap my head around how that translates to my problem. Does anyone have examples where something like this is done?

**Reply by AlCalzone** [reply]
I think you answered my question a while back here: https://spectrum.chat/statecharts/general/xstate-test-with-usemachine-hook~dd820853-dc78-4ecf-a9e9-8331e543602d?m=MTU4NTQxMzkxNzk2Mw==
